File: Game/Scripts/Source/NodeFunctions/Transition.lua
local GlobalCore = require("GlobalCore")
local TransitionCondition = require("Source.NodeFunctions.TransitionCondition")
local FrozenCondition = require("Source.NodeFunctions.FrozenCondition")

local NativeTransition = GlobalCore.Transition
local Transition = {}

function Transition.FreezeTransitionBackground()
    NativeTransition.freezeTransitionBackground()
    return FrozenCondition.new()
end

function Transition.RequestTransition(transitionName, transitionTime)
    transitionName = transitionName == nil and "" or transitionName
    transitionTime = transitionTime == nil and 1.0 or transitionTime
    NativeTransition.requestTransition(transitionName, transitionTime)
    return TransitionCondition.new()
end

return Transition

File: Game/Engine/Source/Global/src/Transition.cpp
#include <Transition.hpp>
#include "System/FramePipelineImpl.hpp"
#include <utility>

void Transition::setTransition(
    const std::shared_ptr<sf::Texture>& transitionResource,
    float transitionTime) {
    ludork::global::system_impl::framePipelineImpl().transition().setTransition(
        transitionResource, transitionTime);
}

void Transition::freezeTransitionBackground() {
    ludork::global::system_impl::framePipelineImpl()
        .transition()
        .freezeTransitionBackground();
}

bool Transition::isTransitionBackgroundFrozen() {
    return ludork::global::system_impl::framePipelineImpl()
        .transition()
        .isTransitionBackgroundFrozen();
}

bool Transition::isTransitionBackgroundFreezePending() {
    return ludork::global::system_impl::framePipelineImpl()
        .transition()
        .isTransitionBackgroundFreezePending();
}

void Transition::cancelTransitionBackgroundFreeze() {
    ludork::global::system_impl::framePipelineImpl()
        .transition()
        .cancelTransitionBackgroundFreeze();
}

void Transition::requestTransition(std::optional<std::string> transitionName,
                                   float transitionTime) {
    ludork::global::system_impl::framePipelineImpl()
        .transition()
        .requestTransition(std::move(transitionName), transitionTime);
}

void Transition::cancelPendingTransition() {
    ludork::global::system_impl::framePipelineImpl()
        .transition()
        .cancelPendingTransition();
}

bool Transition::isTransitionPending() {
    return ludork::global::system_impl::framePipelineImpl()
        .transition()
        .isTransitionPending();
}

bool Transition::isInTransition() {
    return ludork::global::system_impl::framePipelineImpl()
        .transition()
        .isInTransition();
}

void Transition::applyPendingTransition() {
    ludork::global::system_impl::framePipelineImpl()
        .transition()
        .applyPendingTransition();
}
